### DISCUSSION TITLE
New Demo: Dense Wavelength-Division Multiplexing

### DISCUSSION THREAD

**Discussion by kumilingus** [clientIO/joint#2163] [News]
Explore this [network graphical view](https://resources.jointjs.com/demos/joint/demo/dwdm/) used in **fabric-optic communications**.

More specifically, it's a **DWDM Circuit (Dense wavelength-division multiplexing)** which is a method of transmitting multiple optical signals over a single optical fiber by using different wavelengths of light. In a DWDM circuit, each optical signal is assigned a different wavelength _(i.e. color)_, which allows multiple signals to be transmitted simultaneously without interfering with each other.

A graphical representation of the circuit makes it easier for the operator to **monitor** or perform operations on the circuit, and aids in maintenance (e.g. when a link is disconnected or a card fails).

<img width="1516" alt="image" src="https://user-images.githubusercontent.com/3967880/235872219-fa99e8d3-f624-4c2e-bce2-96d09e361406.png">

The DWDM circuit consists of several components, including optical **transceivers**, **multiplexers**, and **demultiplexers**. The optical transceivers are responsible for converting electrical signals into optical signals, and vice versa. The multiplexers are used to combine multiple optical signals into a single optical fiber, while the demultiplexers are used to separate the signals at the other end of the fiber.

The key advantage of a DWDM circuit is that it allows for the efficient use of fiber optic bandwidth. By transmitting multiple signals at different wavelengths, a single optical fiber can carry many more signals than it would be able to use other methods, such as time-division multiplexing.

DWDM circuits are commonly used in long-haul telecommunications networks, where they allow for the efficient transmission of large amounts of data over long distances. They are also used in other applications, such as cable TV and internet service providers, where they allow for the transmission of multiple signals over a single fiber optic connection.

Here is the [source code](https://github.com/clientIO/joint/tree/master/demo/dwdm) for the demo.

**Comment by dkchaitanya92** [comment]
Hi @kumilingus,

When I implemented using JointJs, it is working as expected with Typescript and ReactJs.

I am using JointJs+ in our development.

Above demo, which isn't working as expected when I am implementing with JointJs+.

Below are the files and functionalities are creating the issue for rappid.

**layout.js
----------**
I have replaced the function from "fitToChildren" to "fitEmbeds" as shown below(Not working as expected):

**Code
-----**
=> node.fitToChildren({
            expandOnly: true,
            padding: {
                left: NODE_PADDING_HORIZONTAL,
                right: NODE_PADDING_HORIZONTAL,
                top: NODE_PADDING_VERTICAL + Node.HEADER_HEIGHT,
                bottom: NODE_PADDING_VERTICAL
            },
        }); 


=> node.fitEmbeds({
      padding: {
        left: NODE_PADDING_HORIZONTAL,
        right: NODE_PADDING_HORIZONTAL,
        top: NODE_PADDING_VERTICAL + Node.HEADER_HEIGHT,
        bottom: NODE_PADDING_VERTICAL,
      },
    });

Note: JointJs+ doesn't support "fitToChildren"

**index.ts:
---------**
Facing below issues, when I create "paper" using '@client/rappid' package.

**Code:
-----**
defaultConnector: { name: 'straight', args: { cornerType: 'line', cornerRadius: 6 }},
    defaultRouter: function(_vertices, _opt, linkView) {
        const link = linkView.model;
        const source = link.getSourceCell();
        const target = link.getTargetCell();
        if ((source && isCellHidden(source)) || (target && isCellHidden(target))) {
            // Use straight line if either source or target is hidden
            return [];
        }
        return routers.rightAngle([], { margin: 10 }, linkView);
    },

 

Both "defaultconnecttor" and "defaultRouter with rightAngle()" not available in JointJs+.

I assume "defaultAnchor: multiLinkAnchor" also isn't working.

In "resizePaper",  "transformToFitContent" function isn't available in JointJs+.

I have tried using "scaleContentToFit" and it isn't working as expected.

**Code:
------**
paper.transformToFitContent({
          verticalAlign: 'middle',
          contentArea: { x: 0, y: 0, width, height }
      });

paper.scaleContentToFit({
          //verticalAlign: 'middle',
          contentArea: { x: 0, y: 0, width, height }
      }); 


Please find the screenshots below for the issues facing using JointJs+

Issue 1: When all nodes are expanded, Links and anchors are not working as expected. Please find the screenshot below.


![AllNodesExpanded](https://user-images.githubusercontent.com/47445361/237038134-b82d8a1d-880f-4260-a7ed-c647bfcaf3c1.JPG)

Issue 2: . When nodes are collapsed form, Links, externalLinks, anchors and positions are not properly aligned. Please find the screenshot below.

![expandandcollapsed](https://user-images.githubusercontent.com/47445361/237039519-83ffb693-510b-48a9-bb3e-fe11239a044b.JPG)

Issue 3: Unable to understand why the Paper is positioned at the centre of the Canvas(Client/Chrome screen). Please find the screenshot below. Can you please provide any suggestions to make the paper to start from the beginning/starting position of the Canvas (Client screen).

![paper](https://user-images.githubusercontent.com/47445361/237040354-80cff028-cf19-4b45-a0f3-fe8e64431c92.JPG).

Thanks in Advance.

**Reply by kumilingus** [reply]
Please wait for the JointJS+ (`v3.6`) to catch up with JointJS (`v3.7`) and re-test.
